File: leetcode/155.最小栈/155.最小栈.py
class MinStack:

    def __init__(self):
        self.stk = []
        # 最小值为栈顶，保持栈为降序，push和pop值后栈顶保持可以最小值
        self.min_stk = []

    def push(self, val: int) -> None:
        self.stk.append(val)
        # 新值 <= 栈顶值(最小值)，保持栈为降序
        if not self.min_stk or val <= self.min_stk[-1]:
            self.min_stk.append(val)

    def pop(self) -> None:
        val = self.stk.pop()
        # 如果值是最小值栈顶，也要pop出来，保持数据同步和降序要求
        if val == self.min_stk[-1]:
            self.min_stk.pop()

    def top(self) -> int:
        return self.stk[-1]

    def getMin(self) -> int:
        return self.min_stk[-1]

File: leetcode/155.最小栈/155.最小栈.go
package main

type MinStack struct {
	stk    []int
	minStk []int
}

func Constructor() MinStack {
	return MinStack{stk: []int{},
		minStk: []int{}}
}

func (this *MinStack) Push(val int) {
	this.stk = append(this.stk, val)
	if len(this.minStk) == 0 || val <= this.minStk[len(this.minStk)-1] {
		this.minStk = append(this.minStk, val)
	}
}

func (this *MinStack) Pop() {
	val := this.stk[len(this.stk)-1]
	this.stk = this.stk[:len(this.stk)-1]
	if val == this.minStk[len(this.minStk)-1] {
		this.minStk = this.minStk[:len(this.minStk)-1]
	}
}

func (this *MinStack) Top() int {
	return this.stk[len(this.stk)-1]
}

func (this *MinStack) GetMin() int {
	return this.minStk[len(this.minStk)-1]
}
